File: MyMiniProject/vue-dotnet-finance/backend/Controllers/DashboardController.cs
using System.Security.Claims;
using FinanceApi.Data;
using FinanceApi.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace FinanceApi.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class DashboardController(AppDbContext db) : ControllerBase
{
    private int CurrentUserId => int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
    private bool IsAdmin => User.IsInRole("admin");

    [HttpGet]
    public async Task<ActionResult<DashboardSummary>> Get([FromQuery] int? userId)
    {
        var q = db.Transactions.Include(t => t.User).AsQueryable();

        if (!IsAdmin)
            q = q.Where(t => t.UserId == CurrentUserId);
        else if (userId is not null)
            q = q.Where(t => t.UserId == userId);

        var all = await q.ToListAsync();

        var income = all.Where(t => t.Type == "income").ToList();
        var expense = all.Where(t => t.Type == "expense").ToList();

        var totalIncome = income.Sum(t => t.Amount);
        var totalExpense = expense.Sum(t => t.Amount);

        var byCategory = all
            .GroupBy(t => new { t.Category, t.Type })
            .Select(g => new CategorySum(g.Key.Category, g.Key.Type, g.Sum(x => x.Amount)))
            .OrderByDescending(x => x.Total)
            .ToList();

        var byMonth = all
            .GroupBy(t => t.OccurredAt.ToString("yyyy-MM"))
            .OrderBy(g => g.Key)
            .Select(g => new MonthlySum(
                g.Key,
                g.Where(x => x.Type == "income").Sum(x => x.Amount),
                g.Where(x => x.Type == "expense").Sum(x => x.Amount)))
            .ToList();

        var recent = all
            .OrderByDescending(t => t.OccurredAt)
            .ThenByDescending(t => t.Id)
            .Take(8)
            .Select(t => new TransactionDto(
                t.Id, t.UserId, t.User?.FullName, t.Type, t.Category, t.Title,
                t.Note, t.Amount, t.OccurredAt, t.SlipPath, t.CreatedAt))
            .ToList();

        return Ok(new DashboardSummary(
            totalIncome,
            totalExpense,
            totalIncome - totalExpense,
            income.Count,
            expense.Count,
            byCategory,
            byMonth,
            recent));
    }
}

File: MyMiniProject/laravel-finance/app/Http/Controllers/Api/DashboardController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Transaction;
use App\Models\User;
use Illuminate\Http\Request;

class DashboardController extends Controller
{
    public function show(Request $request)
    {
        /** @var User $auth */
        $auth = $request->attributes->get('auth_user');
        $q = Transaction::with('user');

        if (! $auth->isAdmin()) {
            $q->where('user_id', $auth->id);
        } elseif ($request->filled('userId')) {
            $q->where('user_id', (int) $request->query('userId'));
        }

        $all = $q->get();
        $income = $all->where('type', 'income');
        $expense = $all->where('type', 'expense');

        $totalIncome = (float) $income->sum('amount');
        $totalExpense = (float) $expense->sum('amount');

        $byCategory = $all
            ->groupBy(fn ($t) => $t->category.'|'.$t->type)
            ->map(function ($group) {
                $first = $group->first();

                return [
                    'category' => $first->category,
                    'type' => $first->type,
                    'total' => (float) $group->sum('amount'),
                ];
            })
            ->sortByDesc('total')
            ->values()
            ->all();

        $byMonth = $all
            ->groupBy(fn ($t) => optional($t->occurred_at)?->format('Y-m') ?: 'unknown')
            ->sortKeys()
            ->map(function ($group, $month) {
                return [
                    'month' => $month,
                    'income' => (float) $group->where('type', 'income')->sum('amount'),
                    'expense' => (float) $group->where('type', 'expense')->sum('amount'),
                ];
            })
            ->values()
            ->all();

        $recent = $all
            ->sortByDesc(fn ($t) => [$t->occurred_at, $t->id])
            ->take(8)
            ->values()
            ->map->toApiArray()
            ->all();

        return response()->json([
            'totalIncome' => $totalIncome,
            'totalExpense' => $totalExpense,
            'balance' => $totalIncome - $totalExpense,
            'incomeCount' => $income->count(),
            'expenseCount' => $expense->count(),
            'byCategory' => $byCategory,
            'byMonth' => $byMonth,
            'recent' => $recent,
        ]);
    }
}
